stan/math/prim/fun/quantile.hpp

```diff
@@ -0,0 +1,116 @@
+#include <algorithm>
+#include <vector>
+
+namespace stan {
+namespace math {
+
+/**
+ * Return sample quantiles corresponding to the given probabilities.
+ * The smallest observation corresponds to a probability of 0 and the largest to
+ * a probability of 1.
+ *
+ * Implements algorithm 7 from Hyndman, R. J. and Fan, Y.,
+ * Sample quantiles in Statistical Packages (R's default quantile function)
+ *
+ * @tparam T Type of input vector
+ * @param xs Numeric vector whose sample quantiles are wanted
+ * @param p Probability with value between 0 and 1.
+ * @return Sample quantile.
+ * @throw std::invalid_argument If any element of xs is NaN, or size 0.
+ * @throw std::domain_error If p<0 or p>1.
+ */
+template <typename T, require_vector_t<T>* = nullptr,
+          require_vector_vt<std::is_arithmetic, T>* = nullptr>
+inline double quantile(const T& xs, const double p) {
+  check_not_nan("quantile", "xs", xs);
+  check_nonzero_size("quantile", "xs", xs);
+
+  check_not_nan("quantile", "p", p);
+  check_bounded("quantile", "p", p, 0, 1);
+
+  size_t n_sample = xs.size();
+
+  Eigen::VectorXd x = as_array_or_scalar(xs);
+
+  if (n_sample == 1)
+    return x.coeff(0);
+  if (p == 0.)
+    return *std::min_element(x.data(), x.data() + n_sample);
+  if (p == 1.)
+    return *std::max_element(x.data(), x.data() + n_sample);
+
+  double index = (n_sample - 1) * p;
+  size_t lo = std::floor(index);
+  size_t hi = std::ceil(index);
+
+  std::sort(x.data(), x.data() + n_sample);
+
+  double h = index - lo;
+  return (1 - h) * x.coeff(lo) + h * x.coeff(hi);
+}
+
+/**
+ * Return sample quantiles corresponding to the given probabilities.
+ * The smallest observation corresponds to a probability of 0 and the largest to
+ * a probability of 1.
+ *
+ * Implements algorithm 7 from Hyndman, R. J. and Fan, Y.,
+ * Sample quantiles in Statistical Packages (R's default quantile function)
+ *
+ * @tparam T Type of input vector
+ * @tparam Tp Type of probabilities vector
+ * @param xs Numeric vector whose sample quantiles are wanted
+ * @param ps Vector of probability with value between 0 and 1.
+ * @return Sample quantiles, one for each p in ps.
+ * @throw std::invalid_argument If any of the values are NaN or size 0.
+ * @throw std::domain_error If p<0 or p>1 for any p in ps.
+ */
+template <typename T, typename Tp, require_all_vector_t<T, Tp>* = nullptr,
+          require_vector_vt<std::is_arithmetic, T>* = nullptr,
+          require_vector_vt<std::is_arithmetic, Tp>* = nullptr>
+inline std::vector<double> quantile(const T& xs, const Tp& ps) {
+  check_not_nan("quantile", "xs", xs);
+  check_not_nan("quantile", "ps", ps);
+
+  check_nonzero_size("quantile", "xs", xs);
+  check_nonzero_size("quantile", "ps", ps);
+
+  check_bounded("quantile", "ps", ps, 0, 1);
+
+  size_t n_sample = xs.size();
+  size_t n_ps = ps.size();
+
+  Eigen::VectorXd x = as_array_or_scalar(xs);
+  const auto& p = as_array_or_scalar(ps);
+  std::vector<double> ret(n_ps, 0.0);
```

Comment:
I think you can set the return type to `auto` and then base the return on the input here

```suggestion
  plain_type_t<T> ret(n_ps);
  // Fill with zeros but you don't need to
  fill(ret, 0);
```

So then you can have `Eigen::VectorXd` inputs giving back eigen vectors and `std::vector` inputs giving back `std::vectors`

Reply:
I think this was @bbbales2 suggestion - what are the pros/cons of returning T vs. always returning `std::vector`? Or doesn't matter?

Also - is there a difference between `plain_type_t<T> ret(n_ps);` and `T ret(n_ps);`?

Reply:
`T` in this case can be an eigen expression such as `Eigen::Add<Vec, Vec>` so using `plain_type_t<>` gives you Eigen's `PlainObjectBase` aka something like `Eigen::VectorXd`

Reply:
OK, discussed this with @bbbales2 again - returning only std::vector is indeed intentional (and the argument type for ps should also only be a std vector). This is also what's currently exposed at stanc3 (https://github.com/stan-dev/stanc3/pull/870): 
```
quantile(vector, real) => real
quantile(row_vector, real) => real
quantile(array[] real, real) => real
quantile(vector, array[] real) => array[] real
quantile(row_vector, array[] real) => array[] real
quantile(array[] real, array[] real) => array[] real
```
@SteveBronder do you think differently?

Reply:
Ah! Because our rng functions do the same. Yes cool I think that makes sense